test/e2e/operator_test.go

```diff
@@ -3579,6 +3579,36 @@ func waitForDeploymentEnvVar(t *testing.T, cl client.Client, deployment *appsv1.
 	return err
 }
 
+// waitForDeploymentCompleteWithCleanup waits for a deployment to complete its rollout, then waits for the old
+// generation's pods to finish terminating.
+func waitForDeploymentCompleteWithCleanup(t *testing.T, cl client.Client, deploymentName types.NamespacedName, timeout time.Duration) error {
+	t.Helper()
+	deployment := &appsv1.Deployment{}
+	if err := cl.Get(context.TODO(), deploymentName, deployment); err != nil {
+		return fmt.Errorf("failed to get deployment %s: %w", deploymentName.Name, err)
+	}
+
+	if deployment.Generation != deployment.Status.ObservedGeneration {
+		if err := waitForDeploymentComplete(t, cl, deployment, timeout); err != nil {
+			return fmt.Errorf("timed out waiting for deployment %s to complete rollout: %w", deploymentName.Name, err)
+		}
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return fmt.Errorf("deployment %s has invalid spec.selector: %w", deploymentName.Name, err)
+	}
+
+	return wait.PollImmediate(2*time.Second, timeout, func() (bool, error) {
+		pods := &corev1.PodList{}
+		if err := cl.List(context.TODO(), pods, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+			t.Logf("failed to list pods for deployment %q: %v", deploymentName.Name, err)
+			return false, nil
+		}
+		return len(pods.Items) == int(*deployment.Spec.Replicas), nil
```

Comment:
Is deployment.Spec.Replicas always non-nil?

```suggestion
return len(pods.Items) == int(pointer.Int32Deref(deployment.Spec.Replicas, -1)), nil
```

Reply:
I'm not so sure.  `deployment` may always be non-nil, but `deployment.Spec.Replicas` could very well be nil. Am I missing something? I think it needs to be initialized like in `waitForDeploymentComplete` https://github.com/openshift/cluster-ingress-operator/blob/6660fb0f4fc7eac1f3e8153d1d239c36d063c020/test/e2e/operator_test.go#L3536

Reply:
We could change the Spec.Replicas deref to:

```go
return len(pods.Items) == int(pointer.Int32Deref(deployment.Spec.Replicas, -1)), nil
```

Reply:
https://github.com/openshift/cluster-ingress-operator/pull/904/commits/39c6aef16c999bc4e05e948238454439e8c21b3d changes this to use `Int32Deref`.  I kept the explicit `expectedReplicas` variable so that we aren't calling `Int32Deref` repeatedly; I hope that's fine.